main.cpp

```diff
@@ -1,239 +1,239 @@
-        if(id == current_thread_id){
-            printf("infinite loop!\n");
-        }
-    } while (threads[id].complete);
-
-    // Execute that thread.
-    current_thread_id = id;
-    siglongjmp(threads[current_thread_id].env,1);
-}
-
-int uthread_self(){
-    return current_thread_id;
-}
-
-void start(){
-    current_thread_id = 0;
-    siglongjmp(threads[current_thread_id].env,1);
-}
-
-////////////////////////////////////
-//////////////////////////////////// Test Cases
-////////////////////////////////////
-
-void* uthread_test_function(void* arg){
-    assert((long)arg == 10);
-    for(int i=0;i<100;i++){
-        printf("%d\n", i);
-    }
-    return 0;
-}
-void* uthread_yield_test_function(void* arg){
-    for(int i=0;true;i++){
-        if(i % 500000 == 0){
-            uthread_yield();
-        }
-        printf("%d\n", i);
-    }
-    return 0;
-}
-
-void* uthread_self_test_function(void* arg){
-    static int id = 0;
-    assert(uthread_self()==id);
-    id++;
-    return nullptr;
-}
-
-void* f(void * arg){
-    assert((long)arg == 10);
-    printf("argument: %ld\n", (long)arg);
-    int i=0;
-    while(1) {
-        ++i;
-        printf("in f (%d)\n",i);
-        if(i % 5 == 0){
-        }
-        if (i % 3 == 0) {
-            printf("f: switching\n");
-            uthread_yield();
-        }
-        usleep(SECOND);
-    }
-    return 0;
-}
-
-void* g(void * arg){
-    printf("&arg: %p\n", &arg);
-    printf("arg: 0x%lx\n", (long)arg);
-    int i=0;
-    while(1){
-        ++i;
-        printf("in g (%d)\n",i);
-        if (i % 5 == 0) {
-            printf("g: switching\n");
-            uthread_yield();
-        }
-        usleep(SECOND);
-    }
-    return 0;
-}
-
-void test_uthread_create(){
-    uthread_create(uthread_test_function, (void*)10l);
-}
-
-// Create 10 threads, and check that they alternate between eachother
-void test_thread_yield(){
-    uthread_create(uthread_yield_test_function, nullptr);
-    uthread_create(uthread_yield_test_function, nullptr);
-    uthread_create(uthread_yield_test_function, nullptr);
-    uthread_create(uthread_yield_test_function, nullptr);
-    uthread_create(uthread_yield_test_function, nullptr);
-}
-
-void test_thread_self(){
-    uthread_create(uthread_self_test_function, nullptr);
-}
-
-void* yield_wrapper(void* arg){
-    test_thread_yield();
-}
-
-int main(){
-    test_thread_self();
-    test_uthread_create();
-    uthread_create(yield_wrapper, nullptr);
-    start();
-    return 0;
-}
+ // From http://www-users.cselabs.umn.edu/classes/Fall-2017/csci5103/PROJECT/PROJECT1/sigjmp-demo.c. 
+ 
+ #include <assert.h>
+ #include <stdio.h>
+ #include <setjmp.h>
+ #include <signal.h>
+ #include <string.h>
+ #include <unistd.h>
+ #include <stdlib.h>
+ #include <sys/time.h>
+ 
+ #define SECOND 1000000
+ #define STACK_SIZE 4096
+ 
+ typedef void *(*thread_func)(void *);
+ 
+ struct TCB {
+     sigjmp_buf env;
+     thread_func function;
+     void* arg;
+     bool complete;
```

Comment:
Complete should be replaced with waiting, ready - type variables, as you mentioned.